### ISSUE TITLE
Deadlock reproducer

### ISSUE THREAD

**Issue by ChristopherHogan** [hariharan-devarajan/basket#9]
I've been getting deadlock when simultaneously accessing one remote server on `ares-comp-02` with around 16 (or more) clients on `ares-comp-01`. Here's a minimal reproducer, built from the `0.0.3` branch. Unfortunately it doesn't hang every time, but if you run it several times you should see it hang. When it does hang, it always gets stuck [here](https://github.com/hariharan-devarajan/basket/blob/0.0.3/include/basket/communication/rpc_lib.cpp#L142). Hopefully it's just a mistake in my code. Do you see anything wrong?

### hostfile
```
ares-comp-01:16
ares-comp-02:1
```

### server_list
```
ares-comp-02
```

### MPI Info
```bash
$ mpichversion
MPICH Version:          3.3.1
MPICH Release date:     Wed Jun  5 14:57:33 CDT 2019
MPICH Device:           ch4:ofi
MPICH configure:        --prefix=/home/chogan/local --enable-threads=multiple --with-ibverbs --with-device=ch4:ofi --with-libfabric=/home/chogan/local
MPICH CC:       gcc -I/home/chogan/local/include   -O2
MPICH CXX:      g++   -O2
MPICH F77:      gfortran   -O2
MPICH FC:       gfortran   -O2
```

### repro.cpp
```c++
#include <ostream>
#include <unordered_map>
#include <mpi.h>
#include <basket/unordered_map/unordered_map.h>

struct KeyType{
    size_t a;
    KeyType():a(0){}
    KeyType(size_t a_):a(a_){}
#ifdef BASKET_ENABLE_RPCLIB
    MSGPACK_DEFINE(a);
#endif
    bool operator==(const KeyType &o) const {
        return a == o.a;
    }
    KeyType& operator=( const KeyType& other ) {
        a = other.a;
        return *this;
    }
    bool operator<(const KeyType &o) const {
        return a < o.a;
    }
    bool Contains(const KeyType &o) const {
        return a==o.a;
    }
    friend std::ostream &operator<<(std::ostream &os, const KeyType &k);
#if defined(BASKET_ENABLE_THALLIUM_TCP) || defined(BASKET_ENABLE_THALLIUM_ROCE)
    template<typename A>
    void serialize(A& ar) const {
        ar & a;
    }
#endif
};

std::ostream &operator<<(std::ostream &os, const KeyType &k) {
    os << k.a;
    return os;
}

std::ostream &operator<<(std::ostream &os, std::array<int, 1000> &arr) {
    os << "array<int, 1000>";
    return os;
}

namespace std {
    template<>
    struct hash<KeyType> {
        size_t operator()(const KeyType &k) const {
            return k.a;
        }
    };
}

int main (int argc,char* argv[]) {
    int provided;
    MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
    if (provided < MPI_THREAD_MULTIPLE) {
        printf("Didn't receive appropriate MPI threading specification\n");
        return 1;
    }

    int comm_size;
    int my_rank;
    MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

    bool is_server = my_rank == comm_size - 1;
    bool my_server = 0;

    BASKET_CONF->IS_SERVER = is_server;
    BASKET_CONF->MY_SERVER = my_server;
    BASKET_CONF->NUM_SERVERS = 1;
    BASKET_CONF->SERVER_ON_NODE = is_server;
    BASKET_CONF->SERVER_LIST_PATH = "./server_list";

    const int array_size = TEST_REQUEST_SIZE;
    std::array<int, array_size> my_vals = {0};

    CharStruct shmem_name("CHOGAN_DEADLOCK");
    basket::unordered_map<KeyType, std::array<int, array_size>> *basket_map;
    if (is_server) {
        basket_map = new basket::unordered_map<KeyType, std::array<int, array_size>>(shmem_name);
    }
    MPI_Barrier(MPI_COMM_WORLD);
    if (!is_server) {
        basket_map = new basket::unordered_map<KeyType, std::array<int, array_size>>(shmem_name);
    }

    if (!is_server) {
        int num_requests = 500;
        for(int i = 0; i < num_requests; i++) {
            auto key = KeyType(i);
            basket_map->Put(key, my_vals);
        }
    }

    MPI_Barrier(MPI_COMM_WORLD);
    delete basket_map;
    MPI_Finalize();
    return 0;
}
```

### Makefile
```Makefile
MPICXX ?=mpicxx
CPPFLAGS ?=-DBASKET_ENABLE_RPCLIB -DBASKET_TRACE
CXXFLAGS ?=-I/home/chogan/local/include -std=gnu++1z -g -O3
LDFLAGS ?=-L/home/chogan/local/lib -L/home/chogan/local/lib64 -Wl,-rpath,/home/chogan/local/lib:/home/chogan/local/lib64
LIBS ?=-lpthread -lrt -lmpi -lrpc -lbasket

all: repro

repro: repro.cpp
        $(MPICXX) $(CPPFLAGS) $(CXXFLAGS) $(LDFLAGS) -o $@ $< $(LIBS)
```

### Run
```bash
salloc -N 2 -n 17 --nodelist=ares-comp-[01-02] mpirun -f hostfile ./repro
```
### I build libbasket like this
```bash
pushd basket
if [[ -d build ]]; then
    rm -rf build
fi
mkdir build
pushd build

CXXFLAGS=-I$HOME/local/include                  \
LDFLAGS=-L$HOME/local/lib                       \
    cmake                                       \
        -DCMAKE_INSTALL_PREFIX=$HOME/local      \
        -DCMAKE_BUILD_RPATH=$HOME/local/lib     \
        -DCMAKE_BUILD_TYPE=Release              \
        -DCMAKE_INSTALL_RPATH=$HOME/local/lib   \
        -DCMAKE_CXX_COMPILER=`which mpicxx`     \
        -DCMAKE_C_COMPILER=`which mpicc`        \
        -DBASKET_TRACE=ON                       \
    ..

cmake --build . -- -j 6 VERBOSE=1 && make install
```

**Comment by ChristopherHogan** [comment]
I think I've figured this out. I created a reproducer using just rpclib without libbasket. Then I compiled rpclib with `-DRPCLIB_ENABLE_LOGGING=ON`.  That showed me that the client was failing to connect with the error `asio.system:99`. System errno 99 means failure of a socket to bind because it could not assign the requested address. Some googling said it could be because we're exhausting the port range (`cat /proc/sys/net/ipv4/ip_local_port_range`), and that turning on `SO_REUSEADDR` on sockets before connecting to them would help.  This [patch](https://github.com/qchateau/rpclib/commit/a068800bca7b388bad501e829f751b42c0f941c7) in the rpclib fork does exactly that, which is why I can't reproduce the issue with the fork. This may have security implications though. 
@hariharan-devarajan

**Comment by hariharan-devarajan** [comment]
On large scale systems, especially with KNL processors with 256 cores, we would need this enabled. I would recommend let's update the readme to enable this fix on the original repo. Or update the README to use the forked repo. @ChristopherHogan finalize either of the fix and send a pull request for this issue.

**Comment by ChristopherHogan** [comment]
Fixed by #13.
